### PR TITLE
Enable ARK Operators in PyTorch Models

### PR DESCRIPTION
**Pull request by naturalcandy**
* Integrate ARK modules as custom PyTorch autograd functions.
* Enables compatibility with PyTorch optimizers.

### CODE REVIEW

File: python/ark/module.py [hunk 1]
```diff
@@ -191,3 +202,74 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
...
+            output = grad_input.get_torch_view().clone()
+            for _, param in params_dict.items():
+                if param.staged_tensor is not None:
+                    pytorch_grad = param.staged_tensor.get_torch_view().clone()
+                    param.torch_param.grad = pytorch_grad
+        ctx.ark_module.saved_tensors.clear()
+        return (output, None)
+
+
+class ARKComponent(TorchModule):
```
Comment (chhwang) [inline python/ark/module.py:263]: Let's make `ark.Module` inherit `TorchModule` instead of defining another class.

File: examples/tutorial/model_test_tutorial.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+            grad_output, self.saved_input, transpose_input=True
+        )
+        grad_input = ark.matmul(grad_output, self.weight, transpose_other=False)
+        # Update the gradient of the weight.
+        self.weight.update_gradient(grad_weight)
+        return grad_input, grad_weight
+
+
+# Let's use our previous module to define a double linear layer.
+class MyARKModule(ark.Module):
```
Comment (chhwang) [inline examples/tutorial/model_test_tutorial.py:33]: Let's rename this user interface as `ark.Function` instead of `ark.Module`.

File: python/ark/module.py [hunk 3]
```diff
@@ -191,3 +202,74 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
 
             rt.run()
             return _recursive_ark_to_torch(self.forward_output)
+
+
+class ModuleFn(Function):
```
Comment (chhwang) [inline python/ark/module.py:207]: Please rename this class to something starting with an underscore, as this isn't supposed to be called by users.

**Review by chhwang — CHANGES_REQUESTED** [review]

File: examples/tutorial/model_test_tutorial.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+# Ensure both models have the same weights.
+ark_model.load_state_dict(pytorch_model.state_dict())
+ark_model = replace_layers_with_ark(ark_model)
+
+# Move both models to GPU.
+pytorch_model.to("cuda:0")
+ark_model.to("cuda:0")
+
+# Now let's run the models on some random input.
```
Comment (yzygitzh) [inline examples/tutorial/model_test_tutorial.py:128]: Let's fix random seed to make results reproducible.

File: examples/tutorial/model_test_tutorial.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+    assert torch.allclose(pytorch_output, ark_output, atol=1e-4, rtol=1e-2)
+
+    # Compute losses.
+    torch_loss = loss_fn(pytorch_output, target)
+    ark_loss = loss_fn(ark_output, target)
+
+    # See how ARK's loss compares to PyTorch's loss.
+    print(f"\nPyTorch loss: {torch_loss.item()}")
+    print(f"\nARK loss: {ark_loss.item()}\n")
```
Comment (yzygitzh) [inline examples/tutorial/model_test_tutorial.py:158]: Shall we assert allclose for loss as well?

File: python/ark/tensor.py [hunk 6]
```diff
@@ -241,9 +241,40 @@ class Parameter(Tensor):
...
         Initializes a new instance of the Parameter class.
         """
-        super().__init__(_tensor)
+        if not _no_torch and isinstance(tensor, torch.nn.Parameter):
+            ark_tensor = Tensor.from_torch(tensor)
+            core_tensor = ark_tensor._tensor
+            self.torch_param = tensor
+            self.staged_tensor = None
+        elif isinstance(tensor, _Tensor):
+            core_tensor = tensor
```
Comment (yzygitzh) [inline python/ark/tensor.py:256]: We can keep super().__init__ here

File: python/ark/module.py [hunk 7]
```diff
@@ -191,3 +202,74 @@ def forward(self, *args: Any, **kwargs: Any) -> Any:
...
+        """
+        Converts the gradient outputs to ARK format, computes the gradients for the input
+        and parameters using the ARK module backwards pass, and updates the gradients of the corresponding
+        PyTorch parameters.
+        """
+        ark.init()
+        for name, tensor in ctx.ark_module.saved_tensors.items():
+            if tensor is None or not isinstance(tensor, torch.Tensor):
+                raise ValueError(f"Saved tensor {name} is invalid.")
+            ctx.ark_module.saved_tensors[name] = Tensor.from_torch(tensor)
```
Comment (yzygitzh) [inline python/ark/module.py:247]: Why for saved tensors we store them as PyTorch form and convert them back to ARK form? Why not store ARK form directly
